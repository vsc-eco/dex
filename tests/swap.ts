import {
  logs,
  contract,
  reset,
  stateCache,
  contractEnv,
  setContractImport,
  finalizeTransaction,
  balanceSnapshots,
  contract_id,
  intents,
  ledgerStack,
} from "@vsc.eco/contract-testing-utils";

import { expect as chaiExpect, use, Assertion } from "chai";
import chaiJestSnapshot from "chai-jest-snapshot";
const beforeAll = globalThis.beforeAll || globalThis.before;
use(chaiJestSnapshot);
const expect = chaiExpect as unknown as (
  val: any,
  message?: string
) => typeof Assertion & {
  to: InstanceType<typeof Assertion> & {
    matchSnapshot(): void;
  };
};
// const { beforeAll, beforeEach, describe, it } = globalThis;

const contractImport = import("../build/debug");

beforeAll(() => {
  chaiJestSnapshot.resetSnapshotRegistry();
  return setContractImport(contractImport);
});

beforeEach(async () => {
  // if (globalThis.alert) {
  //   chaiJestSnapshot.configureUsingMochaContext(
  //     // @ts-ignore
  //     (await import("mocha")).default
  //   );
  // }
  return reset();
});

describe("swap", () => {
  it("should parse args", () => {
    expect(
      contract.swapParseArgs(
        JSON.stringify({
          hbd: 10,
          maxSlippage: {
            numerator: "2",
            denominator: "100",
          },
        })
      )
    ).to.eql({
      hbd: 10n,
      hive: 0n,
      withdraw: false,
      maxSlippage: {
        numerator: "2",
        denominator: "100",
      },
    });
  });
  it("should execute", () => {
    balanceSnapshots.set(contract_id, {
      account: contract_id,
      tokens: {
        HIVE: 10000,
        HBD: 3000,
      },
    });
    contractEnv["msg.sender"] = "someAccountName";
    balanceSnapshots.set(contractEnv["msg.sender"], {
      account: contractEnv["msg.sender"],
      tokens: {
        HIVE: 35,
        HBD: 10,
      },
    });
    intents.push({
      name: "hive.allow_transfer",
      args: {
        token: "hbd",
        limit: 10,
      },
    });
    expect(() =>
      contract.swapExec({
        hbd: 10n,
        hive: 0n,
        withdraw: false,
        maxSlippage: {
          numerator: "2",
          denominator: "100",
        },
      })
    ).to.not.throw();
    finalizeTransaction();
  });
  it("should run e2e", () => {
    balanceSnapshots.set(contract_id, {
      account: contract_id,
      tokens: {
        HIVE: 1000000000,
        HBD: 300000000,
      },
    });
    contractEnv["msg.sender"] = "someAccountName";
    balanceSnapshots.set(contractEnv["msg.sender"], {
      account: contractEnv["msg.sender"],
      tokens: {
        HIVE: 35,
        HBD: 10,
      },
    });
    intents.push({
      name: "hive.allow_transfer",
      args: {
        token: "hbd",
        limit: 10,
      },
    });
    expect(
      contract.swap(
        JSON.stringify({
          hbd: 10,
          maxSlippage: {
            numerator: "2",
            denominator: "100",
          },
        })
      )
    ).to.equal(
      JSON.stringify({
        ret: JSON.stringify({
          hive: 33,
          hbd: 10,
          in: "HBD",
          out: "HIVE",
          withdraw: false,
        }),
      })
    );
    finalizeTransaction();
    expect(ledgerStack).to.matchSnapshot();
  });
});
