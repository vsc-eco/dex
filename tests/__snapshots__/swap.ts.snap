// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swap should run e2e 1`] = `
[
  {
    "amount": -10,
    "owner": "someAccountName",
    "to": "vs41qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq99048g",
    "unit": "HBD",
  },
  {
    "amount": 10,
    "from": "someAccountName",
    "owner": "vs41qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq99048g",
    "unit": "HBD",
  },
  {
    "amount": -33,
    "owner": "vs41qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq99048g",
    "to": "someAccountName",
    "unit": "HIVE",
  },
  {
    "amount": 33,
    "from": "vs41qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq99048g",
    "owner": "someAccountName",
    "unit": "HIVE",
  },
]
`;
